src/shared/ui/picker.ts

```diff
@@ -145,3 +148,210 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+                            .next()
+                            .then(newItems => {
+                                getLogger().debug(`Returning a payload of size: ${newItems.value.length}`)
+                                resolve(newItems)
+                            })
+                            .catch(e => {
+                                // append error node
+                                // give quickpick item an error message
+                                // we should not blow away the existing items, they should still be viable
+                                const err = e as Error
+                                getLogger().error('Error while loading items for IteratingQuickPickController:', err)
+                                resolve({
+                                    value: [
+                                        {
+                                            ...IteratingQuickPickController.ERROR_ITEM,
+                                            detail: err.message,
+                                        },
+                                    ],
+                                    done: true,
+                                })
+                            })
+                    }),
+                    // externally-rejectable promise for cancelling an execution
+                    new Promise<IteratorResult<vscode.QuickPickItem[], any>>((resolve, reject) => {
+                        this.state.cancelExecutionFn = () => {
+                            getLogger().debug('Cancelling execution...')
+                            reject()
+                        }
+                    }),
+                ])
+                if (!loadedItems) {
+                    this.state.isDone = true
+                    break
+                }
+                // TODO: Is there a way to append to this ReadOnlyArray so it doesn't constantly pop focus back to the top?
+                this.quickPick.items = this.quickPick.items.concat(loadedItems.value)
+                if (loadedItems.done) {
+                    // no items in response
+                    if (this.quickPick.items.length === 0) {
+                        this.quickPick.items = [IteratingQuickPickController.NO_ITEMS_ITEM]
+                    }
+
+                    this.state.isDone = true
+                    break
+                }
+            } catch (rejectedCallback) {
+                getLogger().debug('Cancelled loop execution')
+                break
+            } finally {
+                // reset cancellation function
+                if (this.state.cancelExecutionFn) {
+                    this.state.cancelExecutionFn = undefined
+                }
+            }
+        }
+
+        // disable loading bar
+        this.quickPick.busy = false
+    }
+}
+
+class IteratingQuickPickControllerState {
+    public cancelExecutionFn: undefined | (() => void) = undefined
+    public isDone: boolean = false
+    public isPaused: boolean = true
```

Comment:
Subtle thing (not a blocker): it's often nice to have default values for booleans be `false`. This works well with interfaces, where you can make the boolean optional and have the default value be `undefined`.

`if (optionalBoolean)` works if `optionalBoolean` is false or undefined.

It also works really well in APIs for the same reason. The absence of a flag is the same as turning that flag off.

That would require you to change `isPaused` to something like `isRunning`

Reply:
It makes sense to me to have the same Boolean default values and to reword the variable